TechnicalDocuments/Proposals/BabylonDesignLibrary.md

```diff
@@ -0,0 +1,136 @@
+The benefit of having a design library is
+1. to develop new UIs faster.
+2. to catch inconsistencies in the UI due to a deisgn flaw or due to an engineering mistake.
+
+The requirement is to cover most of the common use cases and *not* to extend to edge cases.
+Also we must reuse our current tools like `Bento`.
+
+For instance, in our codebase a common UI element is a rounded button with background color. 
+For that we use `Component.Button` and a stylesheet while this approach is functional is has a few drawbacks.
+The attributes of the button (background color, alpha, corner radius etc) can either 
+1. have been redefined in another part of the codebase which is good from a design point of view *and if* we search
+for them we can use them, *if we don't* search for them we will just reimplement something which already exists.
+2. or the attributes of the button haven't been redefined in another part of the application and we may not
+notice that so a design flaw will be implemented. We are humans after all :).
+
+
+## Proposed solution
+We can introduce `AppearenceService` which will be responsible for creating these components following
+the semantics of the design styleguide which will be provided to us.
+
+The purpose of  `AppearenceService` isn't to expose the full potential of `Bento` components but to 
+follow specific design guidelines.
+As an example of this is: 
+A `Bento` component supports different behaviours like  `Deletable` the goal of this
+proposal isn't to replicate them or to make a wrapper around them but to follow specific design guidelines
+and to offer reusable UI elements.
+
+Also with `AppearenceService` we want to limit the exposure of specific components, in order to be able to introduce  
+drastic design changes in the future as easy as possible.
+For example `AppearenceService.Appearence.makeButton` returns `AnyRenderable` instead of `Component.Button`
+because in the future we may decide that we need to use a different button component. We want to depend on abstractions.
+
+Furthermore, `AppearenceService` can also host alerts, custom viewcontroller animations, etc.
+
+``` swift
+final class AppearenceService {
+    let appearence: MutableProperty<Appearence>
+    init(appearence: Appearence) {
+        self.appearence = MutableProperty(appearence)
+    }
+}
+
+extension AppearenceService {
+    struct Appearence {
+        let typography: Typography
+        let colors: Colors
+    }
+}
+
+extension AppearenceService.Appearence {
+    struct Typography {
+        typealias Font = (UIFont, UIColor)
+        let headline1: Font
+        let headline2: Font
+        let body1: Font
+        let body2: Font
+    }
+
+    struct Colors {
+        let primary: UIColor
+        let secondary: UIColor
+    }
+}
+
+extension AppearenceService.Appearence {
+
+     func makeButton(title: String, isEnabled: Bool, didTap: (() -> Void)?) -> AnyRenderable {
```

Comment:
So this will mean we can only have this style of button? What if, in a given renderer, you want to modify the style? Will you have to do it the existing way?

Reply:
@mluisbrown I think the idea is that you will not do it on demand but instead will introduce a new design component if the need for it is agreed by UX team or will use what is there already. So that we don't have dozens of similar but slightly different components.

Reply:
@ilyapuchka tbh I don't know how well that will work with UX and Design. At the very least, we should make the designers aware that we have a fixed set of components.

Reply:
>  we should make the designers aware that we have a fixed set of components.

They are aware of the fact that we *will* have a fixed set of components.
Also as I have been told they don't want to have stuff which doesn't exist in
the arranged fixed set of components (aka `Babylon Design Library`).

If a part of a UI can't be implemented with the `Babylon Design Library` then 
it has to be discussed.